File: ios/Classes/MethodCallHandlerImpl.swift
//
//  MethodCallHandlerImpl.swift
//  opentok_flutter
//
//  Created by Marcelo Rangel on 17/08/20.
//

import Foundation
import OpenTok
import os

public class MethodCallHandlerImpl{
    private let registrar: FlutterPluginRegistrar!
    private var channel: FlutterMethodChannel!
    private var apiKey: String?
    private var sessionId: String?
    private var token: String?
    private var publisherSettings: PublisherSettings?
    private var subscriberSettings: SubscriberSettings?
    public var loggingEnabled: Bool = false
    public var provider: VoIPProvider!

    init(registrar: FlutterPluginRegistrar) {
        self.registrar = registrar
        channel = FlutterMethodChannel(name: "plugins.flutterbr.dev/opentok_flutter", binaryMessenger: registrar.messenger())
        
        channel.setMethodCallHandler {
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            self?.onMethodCall(call: call, result: result)
        }
    }
    
    func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "initialize" {
            guard let args = call.arguments else {
                return
            }

            if let methodArgs = args as? [String: Any] {
                let publisherArg = methodArgs["publisherSettings"] as! String
                let subscriberArg = methodArgs["subscriberSettings"] as? String
                self.loggingEnabled = methodArgs["loggingEnabled"] as? Bool
                
                do {
                    let jsonDecoder = JSONDecoder()

                    self.publisherSettings = try jsonDecoder.decode(PublisherSettings.self, from: publisherArg.data(using: .utf8)!)
                    
                    if (subscriberArg != nil) {
                        self.subscriberSettings = try jsonDecoder.decode(SubscriberSettings.self, from: subscriberArg!.data(using: .utf8)!)
                    }
                    
                } catch {
                    if self.loggingEnabled {
                        print("OpenTok publisher settings error: \(error.localizedDescription)")
                    }
                }
                
                self.apiKey = methodArgs["apiKey"] as? String
                self.sessionId = methodArgs["sessionId"] as? String
                self.token = methodArgs["token"] as? String
                
                provider = VoIPProvider(publisherSettings: self.publisherSettings,
                                        subscriberSettings:self.subscriberSettings,
                                        channel: self,
                                        loggingEnabled: self.loggingEnabled);
                result(true)
            } else {
                result(false)
            }
        } else if call.method == "connect" {
            do {
                try provider?.connect(apiKey: self.apiKey!, sessionId: self.sessionId!, token: self.token!)
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "destroy" {
            do {
                try provider?.disconnect()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "enablePublisherVideo" {
            do {
                try provider?.enablePublisherVideo()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "disablePublisherVideo" {
            do {
                try provider?.disablePublisherVideo()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "unmutePublisherAudio" {
            do {
                try provider?.unmutePublisherAudio()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "mutePublisherAudio" {
            do {
                try provider?.mutePublisherAudio()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "muteSubscriberAudio" {
            do {
                try provider?.muteSubscriberAudio()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "unmuteSubscriberAudio" {
            do {
                try provider?.unmuteSubscriberAudio()
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "switchAudioToSpeaker" {
            do {
                try configureAudioSession(switchedToSpeaker: true)
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "switchAudioToReceiver" {
            do {
                try configureAudioSession(switchedToSpeaker: false)
                result(true)
            } catch {
                result(false)
            }
        } else if call.method == "switchCamera" {
            do {
                try provider.switchCamera()
                result(true)
            } catch {
                result(false)
            }
        } else {
            result(FlutterMethodNotImplemented)
        }
    }

    func channelInvokeMethod(_ method: String, arguments: Any?) {
        channel.invokeMethod(method, arguments: arguments) {
            (result: Any?) -> Void in
            if let error = result as? FlutterError {
                if self.loggingEnabled {
                    if #available(iOS 10.0, *) {
                        os_log("%@ failed: %@", type: .error, method, error.message!)
                    } else {
                        // Fallback on earlier versions
                    }
                }
            } else if FlutterMethodNotImplemented.isEqual(result) {
                if self.loggingEnabled {
                    if #available(iOS 10.0, *) {
                        os_log("%@ not implemented", type: .error)
                    } else {
                        // Fallback on earlier versions
                    }
                }
            }
        }
    }
    
    fileprivate func configureAudioSession(switchedToSpeaker: Bool) throws {
        if self.loggingEnabled {
            print("[FlutterOpenTokViewController] Configure audio session")
            print("[FlutterOpenTokViewController] Switched to speaker = \(switchedToSpeaker)")
        }

        do {
            try AVAudioSession.sharedInstance().setCategory(.playAndRecord, options: [.mixWithOthers, .allowBluetooth])
        } catch {
            if self.loggingEnabled {
                print("[FlutterOpenTokViewController] Session setCategory error: \(error)")
            }
        }

        do {
            try AVAudioSession.sharedInstance().setMode(switchedToSpeaker ? AVAudioSession.Mode.videoChat : AVAudioSession.Mode.voiceChat)
        } catch {
            if self.loggingEnabled {
                print("[FlutterOpenTokViewController] Session setMode error: \(error)")
            }
        }

        do {
            try AVAudioSession.sharedInstance().overrideOutputAudioPort(switchedToSpeaker ? .speaker : .none)
        } catch {
            if self.loggingEnabled {
                print("[FlutterOpenTokViewController] Session overrideOutputAudioPort error: \(error)")
            }
        }

        do {
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            if self.loggingEnabled {
                print("[FlutterOpenTokViewController] Session setActive error: \(error)")
            }
        }
    }
}

File: android/src/main/kotlin/dev/flutterbr/opentok_flutter/MethodCallHandlerImpl.kt
package dev.flutterbr.opentok_flutter

import android.app.Activity
import android.util.Log
import com.opentok.android.AudioDeviceManager
import com.opentok.android.BaseAudioDevice
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.serialization.json.Json

class MethodCallHandlerImpl(private var activity: Activity,
                            messenger: BinaryMessenger?) : MethodChannel.MethodCallHandler {
    private var methodChannel: MethodChannel = MethodChannel(messenger, "plugins.flutterbr.dev/opentok_flutter")
    private var loggingEnabled: Boolean = false
    var provider: VoIPProvider? = null
    private lateinit var publisherSettings: PublisherSettings
    private var subscriberSettings: SubscriberSettings? = null
    private lateinit var apiKey: String
    private lateinit var sessionId: String
    private lateinit var token: String

    init {
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                if (call.arguments == null) {
                    result.success(false)
                    return
                }

                try {
                    val args = call.arguments as? Map<*, *>

                    loggingEnabled = args?.get("loggingEnabled") as Boolean

                    val publisherArg = args["publisherSettings"] as String
                    val subscriberArg = args["subscriberSettings"] as String?
                    apiKey = args["apiKey"] as String
                    sessionId = args["sessionId"] as String
                    token = args["token"] as String

                    try {
                        publisherSettings = publisherArg.let { Json.parse(PublisherSettings.serializer(), it) }
                    } catch (e: Exception) {
                        if (loggingEnabled) {
                            Log.d("[MethodCallHandlerImpl]", "OpenTok publisher settings error: ${e.message}")
                        }
                    }

                    try {
                        if (subscriberArg != null) {
                            subscriberSettings = subscriberArg.let { Json.parse(SubscriberSettings.serializer(), it) }
                        }
                    } catch (e: Exception) {
                        if (loggingEnabled) {
                            Log.d("[MethodCallHandlerImpl]", "OpenTok subscriber settings error: ${e.message}")
                        }
                    }

                    provider = VoIPProvider(activity.applicationContext, publisherSettings, subscriberSettings, this, loggingEnabled)
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }

            "connect" -> {
                try {
                    provider?.connect(apiKey, sessionId, token)
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "disconnect" -> {
                try {
                    provider?.disconnect()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "publishVideo" -> {
                try {
                    provider?.publishVideo()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }

            }
            "unpublishVideo" -> {
                try {
                    provider?.unpublishVideo()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "publishAudio" -> {
                try {
                    provider?.publishAudio()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "unpublishAudio" -> {
                try {
                    provider?.unpublishAudio()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "subscribeToAudio" -> {
                try {
                    provider?.subscribeToAudio()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "unsubscribeToAudio" -> {
                try {
                    provider?.unsubscribeToAudio()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "unsubscribeToVideo" -> {
                try {
                    provider?.unsubscribeToVideo()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "subscribeToVideo" -> {
                try {
                    provider?.subscribeToVideo()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "switchAudioToSpeaker" -> {
                try {
                    configureAudioSession(true)
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "switchAudioToReceiver" -> {
                try {
                    configureAudioSession(false)
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            "switchCamera" -> {
                try {
                    provider?.switchCamera()
                    result.success(true)
                } catch (ex: Exception) {
                    result.success(false)
                }
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun configureAudioSession(switchedToSpeaker: Boolean) {
        if (loggingEnabled) {
            print("[FlutterOpenTokViewController] Configure audio session")
            print("[FlutterOpenTokViewController] Switched to speaker = $switchedToSpeaker")
        }

        if (switchedToSpeaker) {
            AudioDeviceManager.getAudioDevice().outputMode = BaseAudioDevice.OutputMode.SpeakerPhone
        } else {
            AudioDeviceManager.getAudioDevice().outputMode = BaseAudioDevice.OutputMode.Handset
        }
    }


    private fun handleException(exception: java.lang.Exception, result: MethodChannel.Result) {
        channelInvokeMethod("onError", exception)
    }

    fun channelInvokeMethod(method: String, arguments: Any?) {
        methodChannel.invokeMethod(method, arguments, object : MethodChannel.Result {
            override fun notImplemented() {
                if (loggingEnabled) {
                    Log.d("[MethodCallHandlerImpl]", "Method $method is not implemented")
                }
            }

            override fun error(errorCode: String?, errorMessage: String?, errorDetails: Any?) {
                if (loggingEnabled) {
                    Log.d("[MethodCallHandlerImpl]", "Method $method failed with error $errorMessage")
                }
            }

            override fun success(result: Any?) {
                if (loggingEnabled) {
                    Log.d("[MethodCallHandlerImpl]", "Method $method succeeded")
                }
            }

        })
    }

    fun stopListening() {
        methodChannel.setMethodCallHandler(null)
    }

}